### ISSUE TITLE
Discounts: add support for one-time discounts

### ISSUE THREAD

**Issue by michaelbeil** [restrictcontentpro/restrict-content-pro#483]
The current behavior of discounts applies a discount on every recurring payment of a subscription. We should add an option to apply a discount on solely the first payment. 

By default, recurring discounts will be enabled, but this option could be enabled from the Misc tab to disable recurring discounts and enable first payment discounts.

**Comment by derickbailey** [comment]
i'm very much looking forward to this one. right now i have to create a new plan with a negative signup fee, but make everything else match the original plan. having one-time discounts will greatly improve things for me!

**Comment by pippinsplugins** [comment]
I have this started and working for PayPal Standard.

Adding support for the other gateways is a bit trickier. For PayPal Pro and Express, we'll likely need to completely refactor those gateways to mimic the ones in EDD Recurring Payments.

Todo:

- [ ] Stripe
- [ ] PayPal Express
- [ ] PayPal Pro
- [ ] 2Checkout

**Comment by pippinsplugins** [comment]
Needs testing:

- [ ] PayPal Standard
- [ ] 2Checkout
- [ ] PayPal Pro
- [ ] PayPal Express
- [ ] Stripe

**Comment by nosegraze** [comment]
A bit late, I suppose, but I just got a request for this, but also for an option to specify how many times the discount code recurs for (i.e. it applies for 3 months, then returns to full amount). Would be cool. :P

**Comment by pippinsplugins** [comment]
That would be sweet.

**Comment by cklosowski** [comment]
@nosegraze I have something sort of working for that with EDD Recurring and stripe. @pippinsplugins where do we stand on this? Do we need a few gateways updated still or is it basically ready to go? The request @nosegraze got was from someone I know that's looking for this and I'd be happy to jump in and help out.

**Comment by pippinsplugins** [comment]
@cklosowski I still have more work to do on it. I've got it mostly working in 2checkout and paypal standard.

**Comment by pippinsplugins** [comment]
Ok one-time discounts are working for Stripe!

Note: if one-time discounts are enabled, we do _not_ use Stripe coupons. Instead we use the customer balance feature. If one-time discounts are disabled, we use the Stripe coupons.

**Comment by derickbailey** [comment]
SUPER happy to see this work done! this has been a big thing on my list of needs for a while now. :)

i'm curious, though - what's the reason behind the decision for using / not using stripe coupons? stripe supports one-time use discount codes... i'm not familiar with how the "customer balance feature" works, so i'm probably missing something here.

**Comment by pippinsplugins** [comment]
@derickbailey Using the coupons API in Stripe was originally done because the subscriptions API they had back then was _much_ more limited. If we were to rebuild it from scratch today, we wouldn't use the coupons API at all. I chose not to use it here for several reasons:

1. To move us gradually away from the coupons API
2. Prevent potential sync issues. If we used coupons API, we'd have to make an update request to every coupon ever created in RCP to ensure it was still up to date in Stripe

Basically trying to sync coupons in two different databases is a pain and just asking for something to get out of sync.

**Comment by nosegraze** [comment]
Stripe needs a quick fix, which I've noted in the PR. Once that's fixed it works well! :)

**Comment by nosegraze** [comment]
Tested with PayPal Standard - no issues!

**Comment by pippinsplugins** [comment]
Good deal.

@nosegraze could you go ahead and fix the Stripe issue so we can proceed with having that tested?

I'll work on the other gateways ASAP.

**Comment by nosegraze** [comment]
Fixed the Stripe problem. Will give it another test in a bit.

**Comment by mindctrl** [comment]
PayPal Pro logs a payment in RCP at the full amount.

With Stripe, when not recurring, the payment amount added to RCP does not reflect the discounted price paid. It shows the full non-discounted amount. Recurring payments show the correct amount in RCP.

Also with Stripe, a recurring payment results in the `%amount%` not working properly in the email. When not recurring, the `%amount%` tag shows the full non-discounted amount instead of the amount paid.

**Comment by mindctrl** [comment]
Hmm it appears when not recurring, the full amount gets sent to Stripe instead of the discounted amount. This is why the `%amount%` tag showed the wrong price and why the payment amount was wrong in RCP.

**Comment by pippinsplugins** [comment]
@mindctrl I haven't touched PayPal Pro or Express.

**Comment by pippinsplugins** [comment]
@mindctrl I just tested a one-time payment in Stripe and it showed the correct amount for me. Could you try it again?

**Comment by mindctrl** [comment]
@pippinsplugins still happening. I bet your `One Time Discounts` setting was wiped, after switching branches or something. That just happened to me. Once I saved the option again, the incorrect amount behavior showed back up.

**Comment by pippinsplugins** [comment]
@mindctrl yep it was. Will test further.

**Comment by pippinsplugins** [comment]
We have decided to push this back to 2.8, _but_ we are going to make 2.8 a significantly faster release schedule. Hopefully before the end of February.

**Comment by nosegraze** [comment]
Merged `master` into the PR and gave Stripe and PayPal Standard another test - both still worked! I also think I got the issue with Stripe one-time payment working. https://github.com/restrictcontentpro/restrict-content-pro/pull/758/commits/35088444a8045da84d8e75addacf1e4d64d65210

I spent way too long stressing over some funky calculations with proration until I realized I was running into #668. Ugh.

**Comment by pippinsplugins** [comment]
Great! Could we hit PayPal Express next? Once that''s working we can carry it over to PayPal Pro.

**Comment by nosegraze** [comment]
Hah I actually ended up going the other way around. I have Pro done. Haven't been able to test it with recurring yet due to soft descriptor shit, but looks good in theory. I'll push it in a bit.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by nosegraze** [comment]
Updated with PayPal Pro.

Started on PayPal Express. Just needs one thing.. (bear with me, this is my first time touching the PayPal Express API lol)

The initial amount needs to be updated here for the last step: https://github.com/restrictcontentpro/restrict-content-pro/blob/ea926c20809bef2c0cab86c1995be5bc1bbdc128/includes/gateways/class-rcp-payment-gateway-paypal-express.php#L197 But if I'm understanding rightly, there isn't a way to set the `INITAMT` in the first API request in `process_signup()`, sooo I'm not entirely sure how best to get it in `process_confirmation()` where `this->initial_amount` will be `null` due to a page refresh.

**Comment by pippinsplugins** [comment]
We'll need to pass it through the same way we do the other subscription details see #758

**Comment by nosegraze** [comment]
Maybe I'm missing something obvious, but it doesn't look quite that simple.

The initial payment is the amount after discounts and fees are applied. We don't actually save this value in the DB anywhere until after payment is confirmed via webhook. It gets calculated on the registration form and passed to the gateway's constructor, but then the page is refreshed, the gateway class is recreated with no parameters, and all those variables (like initial payment amount) no longer exist. Since we don't save that initial amount in the DB we can't just look it up again. We can get the amount for the subscription and any setup fees, but how do we get the discount code again once that has been wiped after the new page load?

I looked at EDD Recurring a bit and it's simpler there since it looks like that value *is* saved in the database as a pending subscription beforehand, so it can just be looked up again. We don't do that in RCP.

Maybe I'm overlooking something simple? But that's where my thought process is right now.

**Comment by nosegraze** [comment]
You know what.. I just found

```
update_user_meta( $user_data['id'], 'rcp_pending_subscription_amount', $subscription_data['price'] + $subscription_data['fee'] );
```

So never mind? :P

**Comment by nosegraze** [comment]
Updated and ready for testing now.

**Comment by pippinsplugins** [comment]
@mindctrl could you work on testing this today and tomorrow?

**Comment by MikeNGarrett** [comment]
This seems like this will affect #996. I can help test Paypal and Stripe.

**Comment by mindctrl** [comment]
Giving this a test now.

**Comment by mindctrl** [comment]
2Checkout amounts are wrong. When applying a 30% coupon to a $10 sub, I get a $4 transaction in  2CO and a $10 payment entry in RCP.

**Comment by nosegraze** [comment]
https://github.com/restrictcontentpro/restrict-content-pro/commit/743bce106a74afb9c1acde17cb538652f5d8da79 should fix it I think.

**Comment by mindctrl** [comment]
That fixes the RCP payment, but the wrong amount still gets sent to 2CO.

**Comment by nosegraze** [comment]
Right... $4 isn't 30% off. XD I'll take a look.

**Comment by mindctrl** [comment]
Confirmed that fixed 2co issues I was seeing.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mindctrl** [comment]
I'm unable to log into my PayPal sandbox right now to test this. Not sure if it's related to the Amazon outage or not.

**Comment by pippinsplugins** [comment]
I wasn't able to log in earlier either with one account. Tried a different account and it worked.

**Comment by pippinsplugins** [comment]
Just did a test purchase with a discount and Express. It all appeared to work fine, though I can't get sandbox to let me log in to confirm.
